### ISSUE TITLE
Nginx proxy problem "CSRF verification failed"

### ISSUE THREAD

**Issue by reyman** [galexrt/docker-zulip#63]
I use Zulip behind the nginx docker proxy image `jwilder/nginx-proxy `

`2d9c16de8669        jwilder/nginx-proxy                                     "/app/docker-entrypoi"   2 weeks ago         Up 2 hours          0.0.0.0:80->80/tcp, 0.0.0.0:443->443/tcp        nginx`

I configure using the option given in the doc.
```
    VIRTUAL_HOST: "team.comnmodel.org"
    VIRTUAL_PROTO: "https"
    VIRTUAL_PORT: "443"
```
Login page is accessible, but just after authentification i have this, whatever the login/password i test. 

```
Forbidden (403)
CSRF verification failed. Request aborted.

```

So, i stop nginx-proxy, and rebuild image without VIRTUAL_* option

I use directly port 80:80, and 443:443 and the ip of my vps :
https://51.255.47.40/

Error disapears, but i have now a problem to access LDAP, see https://github.com/Galexrt/docker-zulip/issues/62

So, there is some conflict between nginx-proxy and zulip image with self signed certificate?

**Comment by Galexrt** [comment]
@reyman What do you mean by "rebuild image"?

**Comment by reyman** [comment]
I call `rebuild of the image` the `sudo docker-compose up -d` on the last pull. But my view is probably not correct, it's only configuration of global variable i suppose.

**Comment by Galexrt** [comment]
I leave this as a note for me tomorrow (just in case I forget it :D):
@Galexrt Check what headers are forwarded from jwilder's nginx.
```
    proxy_set_header        X-Real_IP       $remote_addr;
    proxy_set_header        X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header        X-NginX-Proxy   true;
    proxy_set_header        Host            $http_host;
    proxy_set_header        Upgrade         $http_upgrade;
    proxy_pass_header       Set-Cookie;
    proxy_pass              http://127.0.0.1:48668/znc/;
```

**Comment by Galexrt** [comment]
@reyman This seems to be a "problem" of the configuration of the jwilder/nginx container. But you should be able to resolve this by looking at this https://github.com/jwilder/nginx-proxy#replacing-default-proxy-settings.

**Comment by Galexrt** [comment]
@reyman Did you try with changing the header configuration in the `jwilder/nginx-proxy`?

**Comment by Galexrt** [comment]
I'm closing this now, as this is a problem with the `jwilder/nginx-proxy` header configuration.

**Comment by reyman** [comment]
@Galexrt Sorry for delay, i try to make ldap works first and test nginx later :)

**Comment by reyman** [comment]
@galexrt Hum i'm not sure to understand here, i need to replace the default nginx config file with which file ? By copying this : https://github.com/jwilder/nginx-proxy#replacing-default-proxy-settings ?

**Comment by galexrt** [comment]
You can try adding the variable `SECURE_PROXY_SSL_HEADER` with value `('HTTP_X_FORWARDED_PROTO', 'https')` to your Zulip `docker-compose.yml`.
I need to make a change to the image first to allow this variable to be set. I'm commenting again when I made this change. You have to use the dev version of the image after I made the change.

(See https://docs.djangoproject.com/en/dev/ref/settings/#secure-proxy-ssl-header)

**Comment by galexrt** [comment]
I pushed the change to the dev image. The image is currently building and should be built in about 20 minutes.

Please try again with the latest `quay.io/galexrt/zulip-postgresql-tsearchextras:latest` and `quay.io/galexrt/zulip:dev`.

**Comment by reyman** [comment]
Ok i made this,
but now i have a `ERR_CONNECTION_REFUSED` for `https://team.comnmodel.org`
Here my docker-compose : https://gist.github.com/cf577cf0801f0c8ed62960862993c3a1

**Comment by galexrt** [comment]
@reyman Could you enter the container and run a `ss -tulpn`.
Entering the container would be `docker exec -it DOCKER_ZULIP /bin/bash`.

**Comment by reyman** [comment]
Here the result : 
```
root@faa3ad619277:/# ss -tulpn                              
Netid  State      Recv-Q Send-Q                                                                    Local Address:Port                                                                      Peer Address:Port 
tcp    LISTEN     0      128                                                                           127.0.0.1:9993                                                                                 *:*     
tcp    LISTEN     0      128                                                                                   *:80                                                                                   *:*      users:(("nginx",99,6))
tcp    LISTEN     0      128                                                                                   *:443                                                                                  *:*      users:(("nginx",99,7))
tcp    LISTEN     0      128                                                                                 ::1:9993                                                                                :::*     
root@faa3ad619277:/# 

```

**Comment by galexrt** [comment]
@reyman You left your `SECRETS_email_password` in. Please remove it!

**Comment by reyman** [comment]
@galexrt sorry about that : https://gist.github.com/03b1523b00d752a2f4de8e71eb61376b

**Comment by galexrt** [comment]
Can you change `SECURE_PROXY_SSL_HEADER` to `SETTING_SECURE_PROXY_SSL_HEADER` and try if it changes something.

**Comment by reyman** [comment]
@galexrt I make the change, but error continue to appear :-/

**Comment by galexrt** [comment]
@reyman The nginx service in the Zulip container is running, so you shouldn't get connection refused.
It seems to caused by Zulip (Django) or jwilder/nginx-proxy.
You could ask at the Zulip project if you need to change Nginx proxy settings to get it working.
I'm sorry, but with that I can't help you. When you create an issue at the Zulip project please mention me so I stay up to date about the problem. Thanks!

**Comment by galexrt** [comment]
@reyman Sorry, for the late response.

Can you comment out the `SECURE_PROXY_SSL_HEADER` environment variable in the `docker-compose.yml`. Enter the Docker Zulip container and run `curl https://127.0.0.1`. You should get the raw html output of the index page or something from Zulip.
Now use `docker inspect --format '{{ .NetworkSettings.IPAddress }}' CONTAINER_NAME_PROXY` to get the Zulip containers IP. Then enter the jwilder nginx proxy container and run `curl https://CONTAINER_IP`, if you get the same ""error" message as the output (see below). If you get the error message only when running `curl ...` inside the jwilder nginx proxy container, it seems to be not a problem of my container or Zulip, but a problem of `jwilder/nginx-proxy`. Maybe there is already an issue filled over at the `jwilder/nginx-proxy`'s project.

***

Error message:
```
Forbidden (403)
CSRF verification failed. Request aborted.
```

**Comment by reyman** [comment]
@galexrt curl is not installed in zulip container, can you push a dev version with curl installed ? :)

**Comment by galexrt** [comment]
@reyman Is this with the `SETTING_SECURE_PROXY_SSL_HEADER` environment variable removed?

**Comment by reyman** [comment]
Yes.
if you want i can test if the double configuration problem i already have
is not the cause.

Le sam. 7 mai 2016 16:41, Alexander Trost <notifications@github.com> a
écrit :

**Comment by galexrt** [comment]
The problem may go away when I have released the new version of the image. See https://github.com/galexrt/docker-zulip/issues/67

**Comment by galexrt** [comment]
Could you try with the `v1.3.12` and see if the error goes away.

**Comment by reyman** [comment]
Sorry for late answer on this thread, I'm on trip, with a very bad wifi
connection :/
I try this next week :)

2016-05-13 12:48 GMT+02:00 Alexander Trost <notifications@github.com>:

> Could you try with the v1.3.12 and see if the error goes away.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/galexrt/docker-zulip/issues/63#issuecomment-219011225>
>



-- 
<http://stackoverflow.com/users/385881/reyman64>

**Comment by galexrt** [comment]
Thanks! Have a nice trip.

**Comment by galexrt** [comment]
**Note**: I'm going to close this issue in about one or two weeks, depending on if there is a new response to this issue.

**Comment by reyman** [comment]
Hi, i'm back, so i can test v1.3.12.
I have one error after docker-compose stop / rm and up : 

```
zulip_1     | Any objects related to these content types by a foreign key will also
zulip_1     | be deleted. Are you sure you want to delete these content types?
zulip_1     | If you're unsure, answer 'no'.
zulip_1     | 
zulip_1     |     Type 'yes' to continue, or 'no' to cancel: Traceback (most recent call last):
zulip_1     |   File "/home/zulip/deployments/current/manage.py", line 25, in <module>
zulip_1     |     execute_from_command_line(sys.argv)
zulip_1     |   File "/usr/lib/python2.7/dist-packages/django/core/management/__init__.py", line 338, in execute_from_command_line
zulip_1     |     utility.execute()
zulip_1     |   File "/usr/lib/python2.7/dist-packages/django/core/management/__init__.py", line 330, in execute
zulip_1     |     self.fetch_command(subcommand).run_from_argv(self.argv)
zulip_1     |   File "/usr/lib/python2.7/dist-packages/django/core/management/base.py", line 393, in run_from_argv
zulip_1     |     self.execute(*args, **cmd_options)
zulip_1     |   File "/usr/lib/python2.7/dist-packages/django/core/management/base.py", line 444, in execute
zulip_1     |     output = self.handle(*args, **options)
zulip_1     |   File "/usr/lib/python2.7/dist-packages/django/core/management/commands/migrate.py", line 225, in handle
zulip_1     |     emit_post_migrate_signal(created_models, self.verbosity, self.interactive, connection.alias)
zulip_1     |   File "/usr/lib/python2.7/dist-packages/django/core/management/sql.py", line 280, in emit_post_migrate_signal
zulip_1     |     using=db)
zulip_1     |   File "/usr/lib/python2.7/dist-packages/django/dispatch/dispatcher.py", line 201, in send
zulip_1     |     response = receiver(signal=self, sender=sender, **named)
zulip_1     |   File "/usr/lib/python2.7/dist-packages/django/contrib/contenttypes/management.py", line 73, in update_contenttypes
zulip_1     |     Type 'yes' to continue, or 'no' to cancel: """ % content_type_display)
zulip_1     | EOFError: EOF when reading a line
zulip_1     | + local RETURN_CODE=0
zulip_1     | + echo 'Zulip migration failed with exit code 0. Exiting.'
zulip_1     | + exit 0
zulip_1     | Zulip migration failed with exit code 0. Exiting.
zulip_zulip_1 exited with code 0
```

**Comment by galexrt** [comment]
@reyman Please open a new issue with the full logs for this new error.

**Comment by reyman** [comment]
Ok, dev image start, i have the `CSRF verification failed. Request aborted.` error, i try to search cause with curl.

**Comment by galexrt** [comment]
Could you please run `curl -k https://172.17.0.9` (with the Zulip container IP) and see if there is still no output.
I'm going to do some more research on this problem on my own, when I find the time for it.

**Comment by reyman** [comment]
@galexrt arg, continue tu see nothing : 

```
➜  ~ sudo docker exec -it zulip_zulip_1 bash                                                              
root@6981c8724ea7:/# curl -k https://172.17.0.9 
root@6981c8724ea7:/# 

```

**Comment by galexrt** [comment]
@reyman Thanks for the output! I'm going to look into the problem on the weekend.

If you have trouble with setting up LDAP that may be related to the image, please keep me informed about them, so I can fix them. :)

**Comment by reyman** [comment]
Some news about this problem @galexrt ?

**Comment by galexrt** [comment]
@reyman I tested without jwilder nginx proxy. Without jwilder nginx proxy it's fully working.
It really seems to be problem with the proxy, but I try to look into why it happens more soon.
Sry about the troubles with this.

**Comment by galexrt** [comment]
@reyman I tried it now with jwilder proxy. I get a redirection loop. What environment variables did you use to start docker-zulip?
Could you please post your current `docker-compose.yml` (without passwords) so I can reproduce the problem?

**Comment by reyman** [comment]
hi @galexrt , sorry for delay, lot of things to do :/
Here you can find my docker-compose : https://gist.github.com/06c81b8fb3b055de2a66574e3e0b2241

**Comment by galexrt** [comment]
Thanks again for your `docker-compose.yml`!
I tested it now with the knowledge from your `docker-compose.yml`.

I use the following env vars and for me it's working (I can login, create streams and write messages, I haven't tested more but with these main features working everything else should work too):
```
VIRTUAL_HOST: "zulip.local.lan"
VIRTUAL_PROTO: "https"
VIRTUAL_PORT: "443"
HTTPS_METHOD: "noredirect"
```
Can you try with this settings? If it's still not working try repulling `jwilder/nginx-proxy` image.
My `jwilder/nginx-proxy` "version" is: Tag `latest` and Image ID `35e54c5e24fa`.

![2016-07-09-091523_941x722_scrot](https://cloud.githubusercontent.com/assets/3718398/16706683/b1455346-45b5-11e6-9f99-6562215e739d.png)

I hope this will finally resolve your issue. :)

**Comment by galexrt** [comment]
I'm closing as I can't reproduce the problem (see comment above it's working for me).
If it's still not working for you, feel free to comment.

***

Thanks for your time helping to create and help to resolve/find a solution for the issue!

**Comment by reyman** [comment]
In parallel i solve some config problem i have with https and jwilder nginx-proxy image.

Thanks for your help, it works now.

**Comment by galexrt** [comment]
@reyman That's good to hear!
Have Fun with docker-zulip and the hopefully soon upcoming official Zulip image I'm working on.

**Comment by MiracleHu** [comment]
proxy_set_header Host $http_host; solve it!!!!
